### PR TITLE
show table in presto doesn't initialized load carbon metadata

### PR DESCRIPTION
**Pull request by photogamerun**
Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Can one of the admins verify this patch?

**Comment by CarbonDataQA** [comment]
Can one of the admins verify this patch?

**Comment by xubo245** [comment]
Please add jira number in title.
Can you provide more infomation?

**Comment by anubhav100** [comment]
Please add the steps to reproduce in PR description @photogamerun

**Comment by ravipesala** [comment]
Can one of the admins verify this patch?

**Comment by CarbonDataQA** [comment]
Can one of the admins verify this patch?

**Comment by chenliang613** [comment]
@photogamerun  please create a jira ticket for this pr and please provide more detail about this pr.

**Comment by photogamerun** [comment]
请看下 Files changed 中的文件的改变(please check the diff in Files changed)

前面的回复是对于第一个方法updateTableList 需要在首先调用updateCarbonFile  方法去初始化carbonFileList这个属性 否则，会出空指针(the above reply is for the first change we need to call updateCarbonFile   before the logic is executed in updateTableList   method )

**Comment by photogamerun** [comment]
第二个问题 change 是这样的
parseCarbonMetadata 这个方法我们发现存在多线程去调用

当时抛了一个这样的异常信息
tableCacheModel.carbonTable should not be null 
后来通过分析代码发现并发模式下，tableCacheModel.carbonTable 中的carbonTable 对象在没有获取前 就已经被另外的线程获取并且读取了。 所以存在读不到表信息的bug

所以在之前实现中有这样一个问题parseCarbonMetadata 中

carbonCache.get().put(table, cache);先调用，

然后走了一个堆逻辑再对cache中的属性做赋值

cache.tableInfo = wrapperTableInfo;
cache.carbonTable = CarbonMetadata.getInstance()
              .getCarbonTable(cache.carbonTableIdentifier.getTableUniqueName());

这样存在一个严重问题，在放入cache 和属性赋值的这个时间段内。
一旦有方法调用carbonCache.get().get(table); 
比如这个方法

 public List<CarbonLocalInputSplit> getInputSplits2(CarbonTableCacheModel tableCacheModel,
                                                     Expression filters, Expression partitionFilters,
                                                     ConnectorSession session)  {
private void removeTableFromCache(SchemaTableName table) {

他们在同时时间都会调用 carbonCache.get().get(table);继而拿取内部属性carbonTable 并调用其api 这个时间其实carbonTable 这个属性是null 调用其api一定会出问题

**Comment by chenliang613** [comment]
add to whitelist

File: integration/presto/src/main/java/org/apache/carbondata/presto/impl/CarbonTableReader.java [hunk 1]
```diff
@@ -314,7 +315,7 @@ private CarbonTable loadTableMetadata(SchemaTableName schemaTableName) {
     for (SchemaTableName table : tableList) {
       if (!table.equals(schemaTableName)) continue;
 
-      return parseCarbonMetadata(table);
+      return parseCarbonMetadata(table).carbonTable;
```
Comment (chenliang613) [inline integration/presto/src/main/java/org/apache/carbondata/presto/impl/CarbonTableReader.java]: why here need to add ".carbonTable"? 

Inside parseCarbonMetadata(table), already have the below code 👍 
result = cache.carbonTable;

File: integration/presto/src/main/java/org/apache/carbondata/presto/impl/CarbonTableReader.java [hunk 2]
```diff
@@ -325,14 +326,23 @@ private CarbonTable loadTableMetadata(SchemaTableName schemaTableName) {
    * @param table name of the given table.
    * @return the CarbonTable instance which contains all the needed metadata for a table.
```
Comment (chenliang613) [inline integration/presto/src/main/java/org/apache/carbondata/presto/impl/CarbonTableReader.java]: please update the comment info also. now is not returning CarbonTable instance.

**Comment by photogamerun** [comment]
please refer to https://github.com/apache/carbondata/pull/2468

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6957/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6961/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5741/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5750/

**Pull request closed without merging**